### ISSUE TITLE
Training using Bisenetv2 with custom dataset of 2 classes

### ISSUE THREAD

**Issue by JJLimmm** [open-mmlab/mmsegmentation#1430]
Hi, if i want to train a model using the bisenetv2 model, and a custom dataset where the classes are either water puddles or background, which files should i be making changes in to ensure that i only want the waterpuddles class to be predicted?

Model: **Bisenetv2**
Dataset: **Custom dataset (ChaseDB1 format)**

Currently, i have made the following changes...

1: Under _configs/_base_/models/bisenetv2.py_, i have changed the number of classes to 2

2: I have created a custom dataset class called _**waterpuddlesDataset()**_, and added into the _mmseg/datasets/waterpuddles.py_ directory. In that file, the classes is defined as **('background', 'waterpuddle')**.

Another question is, how does the **reduce_zero_label** work? 

When i set it to True, my validation accuracy for waterpuddle class become NaN.

**Comment by JJLimmm** [comment]
Sorry, for setting the channel of the output as one, which part of it are you referring to in the base model config file _configs/_base_/models/bisenetv2.py_?

**Comment by JJLimmm** [comment]
> My suggestion is to use BCELoss and set the channel of the output as one.

Sorry, for setting the channel of the output as one, which part of it are you referring to in the base model config file configs/base/models/bisenetv2.py? @MeowZheng

**Comment by MeowZheng** [comment]
num_classes of decode_head

**Comment by JJLimmm** [comment]
> num_classes of decode_head

I changed the channel of the output to 1 and using BCE Loss by setting CrossEntropyLoss's use_sigmoid=True. But i get this error shown below

![error_ss](https://user-images.githubusercontent.com/79782049/160969776-1bef710e-7149-4e62-ba7d-4595b505b453.png)

Below are the config files i used for this run.
Under configs/*
- [configs.zip](https://github.com/open-mmlab/mmsegmentation/files/8385686/configs.zip)


 Under mmseg/datasets/*
- [waterpuddles.zip](https://github.com/open-mmlab/mmsegmentation/files/8385692/waterpuddles.zip)

**Comment by MeowZheng** [comment]
please fetch the master branch, we have fix the bug

**Comment by JJLimmm** [comment]
Hi @MeowZheng , i am still getting the same error after fetching the new fixes. 

![image](https://user-images.githubusercontent.com/79782049/161016301-baaed3bf-5ab3-45e2-bd08-77b70df0e14e.png)

Configs used are still the same as the previous replies above.

**Comment by RockeyCoss** [comment]
you can try seting `num_class=2` .

**Comment by JJLimmm** [comment]
> Hi @JJLimmm There are some problems with our BCELoss currently. You can try setting `num_class=2` and ·`use_sigmoid=False` .

Alright noted! However, for Binary class loss, shouldnt it only be 1 class? 

And if i were to set `num_class=2`  , do i have to set anything else like shown below

- setting `ignore_index=None` in decode_head?
- setting `avg_non_ignore=True`?
- for `reduce_zero_label` in custom dataset?

**Comment by JJLimmm** [comment]
Right now, i have set the following configs and run training once...
- `CLASSES = ('background', 'waterpuddles')`
- `reduce_zero_label=True`
- In decode_head and auxiliary heads for Bisenetv2,
- `num_classes=2`
- `use_sigmoid=False`
- `use_mask=False`
- `avg_non_ignore=True`

And during validation in training run, i got the following output.

![image](https://user-images.githubusercontent.com/79782049/161188396-571c2fde-c1e6-4417-b0a0-ad91660075af.png)

The waterpuddles class is 0 for IoU.

>

**Comment by RockeyCoss** [comment]
Well, if you set `reduce_zero_label=True`, then the class corresponds to `0` in the label, `background` in this case, should not appear in the `CLASSES` tuple. In this 2-class-segmentation problem, setting `reduce_zero_label=True` is equivalent to ignoring all the background pixels, which is not recommended. Because by doing so, your model can only see the pixels of class `waterpuddles` during training. In other words, the model is merely trained on positive samples, which may affect its performance when inferencing an image with no pixels belonging to class `waterpuddles`.
    As to `avg_non_ignore`, it means that the loss is only averaged over non-ignored targets. I recommend setting it `True` since it's more reasonable. Reduce_zero_label defaults to `255`, so you don't need to set it as `None`.
   In summary, here are the settings I recommend:
- `CLASSES = ('background', 'waterpuddles')`
- `reduce_zero_label=False`
- In decode_head and auxiliary heads for Bisenetv2,
- `num_classes=2`
- `use_sigmoid=False`
- `use_mask=False`
- `avg_non_ignore=True`

**Comment by RockeyCoss** [comment]
> Alright noted @RockeyCoss ! However, for Binary class loss, shouldnt it only be 1 class?

Yes, you are right. However, there is a problem with our BCELoss currently. We will fix it as soon as possible. Setting `num_classes=2` and `use_sigmoid=False` is basically the same.

**Comment by JJLimmm** [comment]
> > Alright noted @RockeyCoss ! However, for Binary class loss, shouldnt it only be 1 class?
> 
> Yes, you are right. However, there is a problem with our BCELoss currently. We will fix it as soon as possible. Setting `num_classes=2` and `use_sigmoid=False` is basically the same.

Alright noted, thanks for the help and will await the bug fix for this!

**Comment by MeowZheng** [comment]
we have fixxed it in #1454
